tools/export/exporters.py

```diff
@@ -1,245 +1,220 @@
+                                self.resources.file_basepath[config_header])
+        flags = {key + "_flags": value for key, value
+                 in self.toolchain.flags.iteritems()}
+        if config_header:
+            flags['c_flags'] += self.toolchain.get_config_option(config_header)
+            flags['cxx_flags'] += self.toolchain.get_config_option(
+                config_header)
+        return flags
+
+    def get_source_paths(self):
+        """Returns a list of the directories where source files are contained"""
+        source_keys = ['s_sources', 'c_sources', 'cpp_sources', 'hex_files',
+                       'objects', 'libraries']
+        source_files = []
+        for key in source_keys:
+            source_files.extend(getattr(self.resources, key))
+        return list(set([os.path.dirname(src) for src in source_files]))
 
     def progen_get_project_data(self):
         """ Get ProGen project data  """
         # provide default data, some tools don't require any additional
         # tool specific settings
-        code_files = []
-        for r_type in ['c_sources', 'cpp_sources', 's_sources']:
-            for file in getattr(self.resources, r_type):
-                code_files.append(file)
-
-        sources_files = code_files + self.resources.hex_files + self.resources.objects + \
-            self.resources.libraries
-        sources_grouped = Exporter._get_dir_grouped_files(sources_files)
-        headers_grouped = Exporter._get_dir_grouped_files(self.resources.headers)
-
-        project_data = {
-            'common': {
-                'sources': sources_grouped,
-                'includes': headers_grouped,
-                'build_dir':'.build',
-                'target': [TARGET_MAP[self.target].progen['target']],
-                'macros': self.get_symbols(),
-                'export_dir': [self.inputDir],
-                'linker_file': [self.resources.linker_script],
-            }
-        }
+
+        def make_key(src):
+            """turn a source file into it's group name"""
+            key = os.path.basename(os.path.dirname(src))
+            if not key:
+                key = os.path.basename(os.path.normpath(self.export_dir))
+            return key
+
+        def grouped(sources):
+            """Group the source files by their encompassing directory"""
+            data = sorted(sources, key=make_key)
+            return {k: list(g) for k, g in groupby(data, make_key)}
+
+        if self.toolchain.get_config_header():
+            config_header = self.toolchain.get_config_header()
+            config_header = relpath(config_header,
+                                    self.resources.file_basepath[config_header])
+        else:
+            config_header = None
+
+        # we want to add this to our include dirs
+        config_dir = os.path.dirname(config_header) if config_header else []
+
+        project_data = ProjectTemplateInternal._get_project_template()
```

Comment:
> fixing paths

This is the problematic part. I can revert exporters.py and look at the speed difference. I am sure it will not be much, but I don't want progen to fix these paths. They have already been manipulated within the Resource object. The way progen handles relative paths does not handle this use case.  I have created an issue [here](https://github.com/project-generator/project_generator/issues/419).